### ISSUE TITLE
psc make can't read externs with unicode

### ISSUE THREAD

**Issue by natefaubion** [purescript/purescript#2251]
`psc` always rebuilds about 80% of our source tree even when nothing has changed. I've tracked it down to it failing to decode the externs file for our custom Prelude.

```
Left "Failed reading: Cannot decode byte '\\xd7': Data.Text.Internal.Encoding.decodeUtf8: Invalid UTF-8 stream"
```

`\xd7` is the unicode multiplication sign (we define a bunch of unicode operators in our Prelude).

**Comment by natefaubion** [comment]
We are still on on 0.8 for the moment (hoping to fix that soon), but has reading externs in `Make` changed in 0.9?

**Comment by kRITZCREEK** [comment]
Does this happen without any fast rebuilding involved? Just to make sure the "changing" bit is not psc-ide.

**Comment by natefaubion** [comment]
We almost never change SlamData.Prelude, and we frequently do clean builds, so I'd be surprised.

**Comment by natefaubion** [comment]
FWIW, this happens immediately after a clean build, no psc-ide involved.

**Comment by natefaubion** [comment]
https://gist.github.com/natefaubion/05eab491e1f3792d2e1aa47edd5e554d the externs in question. \xd7 is the first unicode operator encountered right at the top.

**Comment by natefaubion** [comment]
Related? http://stackoverflow.com/questions/27669418/aeson-does-not-decode-strings-with-unicode-characters

**Comment by garyb** [comment]
I don't think so, the file is read with `readUTF8File` - so unless that uses `pack` internally it should be ok.

**Comment by garyb** [comment]
Err, right, and we wrote `readUTF8File` :smile:  https://github.com/purescript/purescript/blob/3154b245e37868141dee04c6cb1269028df273c4/src/System/IO/UTF8.hs#L14-L18 (no `pack`).

**Comment by natefaubion** [comment]
It uses `fromString` though to convert it from a `String` to a ByteString, which is where its failing?

**Comment by natefaubion** [comment]
Indeed, `fromString` uses the same implementation as `pack` from `Char8`.

**Comment by garyb** [comment]
So we need to pack it into text, and then `encodeUtf8` it into a `ByteString` then. Seems sensible. :laughing:

**Comment by natefaubion** [comment]
To be fair, Text is just an opaque wrapper around ByteString I think, so its not like its expensive, just roundabout.

**Comment by kRITZCREEK** [comment]
maybe https://hackage.haskell.org/package/bytestring-0.10.8.1/docs/Data-ByteString-Builder.html#v:stringUtf8 is enough?

**Comment by natefaubion** [comment]
Indeed it does! With this change it works properly.

**Comment by hdgarrood** [comment]
Isn't Text UTF16 internally? Whatever the encoding is, I don't think it's UTF8, so I think Text -> ByteString is at least O(n) (not that that's necessarily a problem, but I don't think it's free).

I'd much rather avoid `String` entirely here if we can, and I think we should be able to. What about using the `readFile` from `Data.ByteString` instead?

**Comment by natefaubion** [comment]
readExterns is part of the Make API, so I think it's best to leave something like that to when we move to Text.

**Comment by garyb** [comment]
Resolved by #2253
